File: boj/Math/BOJ_20302.swift
/*
 21.10.29
 백준 20302 민트 초코
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let a = readLine()!
let arr = readLine()!.split(separator: " ").map { String($0) }

var prime = [Int](repeating: 0, count: 100001)
var numerator = true

func factorization (_ num: Int) {
  let n = Int(sqrt(Double(num))) + 1
  var x = num
  
  for i in 2 ... n {
    while x % i == 0 {
      x /= i
      if numerator { prime[i] += 1 }
      else { prime[i] -= 1 }
    }
  }
  
  if x > 1 {
    if numerator { prime[x] += 1 }
    else { prime[x] -= 1 }
  }
}

func main() {
  for i in 0 ..< arr.count {
    if arr[i] == "0" {
      print("mint chocolate")
      return
    }
    
    if i % 2 == 0 {
      factorization(abs(Int(arr[i])!))
    } else {
      numerator = arr[i] == "*"
    }
  }
  
  for i in 2 ..< prime.count {
    if prime[i] < 0 {
      print("toothpaste")
      return
    }
  }
  
  print("mint chocolate")
}

main()

File: boj/Math/BOJ_20302.cpp
/*
 21.10.29
 백준 20302 민트 초코
 https://github.com/skyqnaqna/algorithm_study
 */

#include <bits/stdc++.h>

using namespace std;
using ll = long long;
using dd = double;
using pii = pair<int, int>;
using pll = pair<ll, ll>;

#define INF (ll)2e12
#define endl "\n"

void input() {
    freopen("input.txt", "r", stdin);
}

int n, prime[100001];
bool numerator;

void factorization(int num) {
    int end = (int)sqrt(num) + 1;
    int x = num;

    for (int i = 2; i <= end; ++i) {
        while (x % i == 0) {
            x /= i;

            if (numerator)  ++prime[i];
            else --prime[i];
        }
    }

    if (x > 1) {
        if (numerator) ++prime[x];
        else --prime[x];
    }
}

int main() {
    ios::sync_with_stdio(0); cin.tie(0);

    cin >> n;

    numerator = true;
    bool zero = false;
    for (int i = 0; i < 2 * n - 1; ++i) {
        if (i % 2 == 0) {
            int x; cin >> x;
            if (x == 0) {
                zero = true;
                continue;
            }

            factorization(abs(x));
        } else {
            string s; cin >> s;

            numerator = s == "*";
        }
    }

    if (zero) {
        cout << "mint chocolate\n";
        return 0;
    }

    for (auto &res : prime) {
        if (res < 0) {
            cout << "toothpaste\n";
            return 0;
        }
    }

    cout << "mint chocolate\n";

    return 0;
}